pkg/controller/envoy/ads_event_test.go

```diff
@@ -0,0 +1,642 @@
+	config_route_v3 "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	service_discovery_v3 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	cluster_v2 "kmesh.net/kmesh/api/v2/cluster"
+	core_v2 "kmesh.net/kmesh/api/v2/core"
+	cache_v2 "kmesh.net/kmesh/pkg/cache/v2"
+	"kmesh.net/kmesh/pkg/utils/hash"
+)
+
+func TestHandleCdsResponse(t *testing.T) {
+	t.Run("new cluster, cluster type is eds", func(t *testing.T) {
+		adsLoader := NewAdsLoader()
+		svc := &ServiceEvent{
+			DynamicLoader: adsLoader,
+		}
+		cluster := &config_cluster_v3.Cluster{
+			Name: "ut-cluster",
+			ClusterDiscoveryType: &config_cluster_v3.Cluster_Type{
+				Type: config_cluster_v3.Cluster_EDS,
+			},
+		}
+		anyCluster, err := anypb.New(cluster)
+		assert.NoError(t, err)
+		rsp := &service_discovery_v3.DiscoveryResponse{
+			Resources: []*anypb.Any{
+				anyCluster,
+			},
+		}
+		err = svc.handleCdsResponse(rsp)
+		assert.NoError(t, err)
+		assert.Equal(t, []string{"ut-cluster"}, svc.DynamicLoader.edsClusterNames)
+		wantHash := hash.Sum64String(anyCluster.String())
+		actualHash := svc.DynamicLoader.ClusterCache.GetCdsHash(cluster.GetName())
+		assert.Equal(t, wantHash, actualHash)
+	})
+
+	t.Run("new cluster, cluster type is not eds", func(t *testing.T) {
+		adsLoader := NewAdsLoader()
+		svc := &ServiceEvent{
+			DynamicLoader: adsLoader,
+		}
+		cluster := &config_cluster_v3.Cluster{
+			Name: "ut-cluster",
+			ClusterDiscoveryType: &config_cluster_v3.Cluster_Type{
+				Type: config_cluster_v3.Cluster_LOGICAL_DNS,
+			},
+		}
+		anyCluster, err := anypb.New(cluster)
+		assert.NoError(t, err)
+		rsp := &service_discovery_v3.DiscoveryResponse{
+			Resources: []*anypb.Any{
+				anyCluster,
+			},
+		}
+		err = svc.handleCdsResponse(rsp)
+		assert.NoError(t, err)
+		assert.Equal(t, []string{}, svc.DynamicLoader.edsClusterNames)
+		wantHash := hash.Sum64String(anyCluster.String())
+		actualHash := svc.DynamicLoader.ClusterCache.GetCdsHash(cluster.GetName())
+		assert.Equal(t, wantHash, actualHash)
+	})
+
+	t.Run("already have cluster, change cluster hash", func(t *testing.T) {
+		adsLoader := NewAdsLoader()
+		svc := &ServiceEvent{
+			DynamicLoader: adsLoader,
+		}
+		cluster := &config_cluster_v3.Cluster{
+			Name: "ut-cluster",
+			ClusterDiscoveryType: &config_cluster_v3.Cluster_Type{
+				Type: config_cluster_v3.Cluster_LOGICAL_DNS,
+			},
+		}
+		anyCluster, err := anypb.New(cluster)
+		assert.NoError(t, err)
+		rsp := &service_discovery_v3.DiscoveryResponse{
+			Resources: []*anypb.Any{
+				anyCluster,
+			},
+		}
+		err = svc.handleCdsResponse(rsp)
+		assert.NoError(t, err)
+
+		cluster.AltStatName = "ut-name"
+		anyCluster, err = anypb.New(cluster)
+		assert.NoError(t, err)
+		rsp = &service_discovery_v3.DiscoveryResponse{
+			Resources: []*anypb.Any{
+				anyCluster,
+			},
+		}
+		err = svc.handleCdsResponse(rsp)
```

Comment:
why donot you need to mock the map r/w  like maps_v2.ClusterUpdatep

Reply:
maps_v2.ClusterUpdate handles maps in the kerne. And I dont use ApiStatu as a basis for testing. So I dont mock maps_v2.ClusterUpdatep.

Reply:
I will add testcase of Flush() in ut of cache/v2/cluster_test.go to cover usecases of handleCdsResponse

Reply:
UT is to simulate real sceanrio, i am not sure how you make it no bpf map write when a new cluster  added